source/code/source/playfab/PlayFabPluginManager.cpp

```diff
@@ -27,47 +29,43 @@ namespace PlayFab
 
     std::shared_ptr<IPlayFabPlugin> PlayFabPluginManager::GetPluginInternal(const PlayFabPluginContract contract, const std::string& instanceName)
     {
-        std::pair<PlayFabPluginContract, std::string> key = std::make_pair(contract, instanceName);
-        const auto pluginEntry = plugins.find(key);
-        if (pluginEntry == plugins.end())
+        PluginEntry& entry = FindOrCreatePluginEntry(contract, instanceName);
+        if (entry.plugin == nullptr)
         {
-            // Requested plugin is not in the cache, create the default one
-            std::shared_ptr<IPlayFabPlugin> pluginPtr = nullptr;
             switch (contract)
             {
             case PlayFabPluginContract::PlayFab_Serializer:
-                pluginPtr = CreatePlayFabSerializerPlugin();
+                entry.plugin = CreatePlayFabSerializerPlugin();
                 break;
             case PlayFabPluginContract::PlayFab_Transport:
-                pluginPtr = CreatePlayFabTransportPlugin();
+                entry.plugin = CreatePlayFabTransportPlugin();
                 break;
             default:
                 throw PlayFabException(PlayFabExceptionCode::PluginAmbiguity, "This contract is not supported");
                 break;
             }
-
-            plugins.insert({ key, pluginPtr });
-            return pluginPtr;
-        }
-        else
-        {
-            return pluginEntry->second;
         }
+        return entry.plugin;
     }
 
     void PlayFabPluginManager::SetPluginInternal(const std::shared_ptr<IPlayFabPlugin>& plugin, const PlayFabPluginContract contract, const std::string& instanceName)
     {
-        std::pair<PlayFabPluginContract, std::string> key = std::make_pair(contract, instanceName);
-        const auto pluginEntry = plugins.find(key);
-        if (pluginEntry == plugins.end())
-        {
-            plugins.insert({ key, std::move(plugin) });
-        }
-        else
+        PluginEntry& entry = FindOrCreatePluginEntry(contract, instanceName);
+        entry.plugin = plugin;
+    }
+
+    PlayFabPluginManager::PluginEntry& PlayFabPluginManager::FindOrCreatePluginEntry(PlayFabPluginContract contract, const std::string& instanceName)
+    {
+        for (PluginEntry& pluginEntry : plugins)
         {
-            plugins.erase(key);
-            plugins.insert({ key, std::move(plugin) });
+            if (pluginEntry.contract == contract &&
+                pluginEntry.name == instanceName)
+            {
+                return pluginEntry;
+            }
         }
+        plugins.push_back({contract, instanceName, nullptr});
```

Comment:
This could be refactored towards emplace_back to save a potential string allocation, but it's not a blocker given the use-case.

Reply:
I don't think emplace_back saves a string allocation.
Given move semantics, I think the only difference between empalce vs push_back is where the string allocation takes place. with push_back, the copy/allocation takes place when the r-value `{contract, instanceName, nullptr}` is constructed. After, the copied string value is moved into the constructed element. with emplace_back, the copy just happens as part of building the constructed element.
That being said, if emplace_back works here, I think it's the better API to use since it's more clear